### PR TITLE
ARROW-6825: [C++] Rework CSV reader IO around readahead iterator

### PR DESCRIPTION
**Pull request by pitrou**
Make the delimiting chunker a common facility used by CSV and JSON.

### CODE REVIEW

**Comment by pitrou** [comment]
@bkietz

**Comment by pitrou** [comment]
@kou There is a weird Github Actions error here: https://github.com/apache/arrow/pull/5727/checks?check_run_id=273513110

File: cpp/src/arrow/util/delimiting.h [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+  virtual ~Delimiter();
+
+  /// \brief Find the position of the first delimiter inside block
+  ///
+  /// `partial` is taken to be the beginning of the block, and `block`
+  /// its continuation.  Also, `partial` doesn't contain a delimiter.
+  ///
+  /// The returned `out_pos` is relative to `block`'s start and should point
+  /// *after* the first delimiter.
+  /// `out_pos` will be -1 if no delimiter is found.
```
Comment (bkietz) [inline cpp/src/arrow/util/delimiting.h]: please add `static constexpr int64_t kNoDelimiterFound = -1;` (static member of Delimiter) to make this reserved value more clear everywhere

File: cpp/src/arrow/json/chunker.cc [hunk 2]
```diff
@@ -187,88 +118,60 @@ static size_t ConsumeWholeObject(Stream&& stream) {
...
+class ParsingDelimiter : public Delimiter {
  public:
-  Status Process(std::shared_ptr<Buffer> block, std::shared_ptr<Buffer>* whole,
-                 std::shared_ptr<Buffer>* partial) override {
-    if (block->size() == 0) {
-      *whole = SliceBuffer(block, 0, 0);
-      *partial = block;
-      return Status::OK();
+  Status FindFirst(string_view partial, string_view block, int64_t* out_pos) override {
+    // XXX should we bubble up JSON parse errors instead of ignoring them?
```
Comment (bkietz) [inline cpp/src/arrow/json/chunker.cc]: I think it's fine to wait; in the (unusual) case of a parse error we'll just fail a little later by deferring those checks to the thread pool

**Review by bkietz — CHANGES_REQUESTED** [review]
Very elegant. Just a few comments

File: cpp/src/arrow/util/delimiting.h [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+  /// \param[in] block json data to be chunked
+  /// \param[out] whole subrange of block containing whole json objects
+  /// \param[out] partial subrange of block a partial json object
+  Status Process(std::shared_ptr<Buffer> block, std::shared_ptr<Buffer>* whole,
+                 std::shared_ptr<Buffer>* partial);
+
+  /// \brief Carve the completion of a partial object out of a block
+  ///
+  /// Pre-conditions:
+  /// - `partial` is the start of a valid block of delimited data
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/util/delimiting.h]: Right.

File: cpp/src/arrow/csv/chunker.cc [hunk 4]
```diff
@@ -18,168 +18,246 @@
...
-  while (size > 0) {
-    if (*s == '\r' || *s == '\n') {
-      return s;
-    }
-    --s;
-    --size;
-  }
-  return nullptr;
-}
+// NOTE: cvsmonkey (https://github.com/dw/csvmonkey) has optimization ideas
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/csv/chunker.cc]: AH :-)

File: cpp/src/arrow/util/delimiting.h [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+};
+
+ARROW_EXPORT
+std::shared_ptr<Delimiter> MakeNewlineDelimiter();
+
+/// \brief A reusable block-based chunker for delimited data
+///
+/// The chunker takes a block of delimited data and finds a suitable place
+/// to cut it up without splitting an object.
+class ARROW_EXPORT DelimitedChunker {
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/util/delimiting.h]: No preference from me.

File: cpp/src/arrow/json/chunker.h [hunk 6]
```diff
@@ -19,76 +19,17 @@
...
-                              std::shared_ptr<Buffer>* completion,
-                              std::shared_ptr<Buffer>* rest) = 0;
-
-  static std::unique_ptr<Chunker> Make(const ParseOptions& options);
-
- protected:
-  Chunker() = default;
-  ARROW_DISALLOW_COPY_AND_ASSIGN(Chunker);
-};
+ARROW_EXPORT
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/json/chunker.h]: `chunker.h` is small, but `chunker.cc` is non-trivial IMHO.

**Comment by pitrou** [comment]
@bkietz I think I've addressed all your comments.
@fsaintjacques Are you still willing to review?

**Review by bkietz — APPROVED** [review]
LGTM

File: cpp/src/arrow/csv/parser_test.cc [hunk 7]
```diff
@@ -360,6 +386,12 @@ TEST(BlockParser, Final) {
   csv = MakeCSVData({"ab,cd"});
   AssertParseFinal(parser, csv);
   AssertColumnsEq(parser, {{"ab"}, {"cd"}});
+
+  // Two blocks
+  auto csv1 = MakeCSVData({"ab,cd\n"});
+  auto csv2 = MakeCSVData({"ef,"});
+  AssertParseFinal(parser, {csv1, csv2});
```
Comment (pprudhvi) [inline cpp/src/arrow/csv/parser_test.cc]: @pitrou I'm getting 'ambiguous function call' compilation errors with this and AssertParseOk(line 241). Not sure why this was not caught in CI builds. I am building with tests, gandiva, jni ON. Could you please take a look and let me know if I should set any compiler flags? Thanks.

Comment (pitrou) [reply cpp/src/arrow/csv/parser_test.cc]: Can you post the full compiler output?

Comment (pprudhvi) [reply cpp/src/arrow/csv/parser_test.cc]: `
/Users/travis/build/dremio/arrow-build/arrow/cpp/src/arrow/csv/parser_test.cc:241:5: error: call to 'AssertParseOk' is ambiguous

    AssertParseOk(parser, {csv1, csv2});

    ^~~~~~~~~~~~~

/Users/travis/build/dremio/arrow-build/arrow/cpp/src/arrow/csv/parser_test.cc:138:6: note: candidate function

void AssertParseOk(BlockParser& parser, const std::string& str) {

     ^

/Users/travis/build/dremio/arrow-build/arrow/cpp/src/arrow/csv/parser_test.cc:144:6: note: candidate function

void AssertParseOk(BlockParser& parser, const std::vector<util::string_view>& data) {

     ^

/Users/travis/build/dremio/arrow-build/arrow/cpp/src/arrow/csv/parser_test.cc:393:3: error: call to 'AssertParseFinal' is ambiguous

  AssertParseFinal(parser, {csv1, csv2});

  ^~~~~~~~~~~~~~~~

/Users/travis/build/dremio/arrow-build/arrow/cpp/src/arrow/csv/parser_test.cc:150:6: note: candidate function

void AssertParseFinal(BlockParser& parser, const std::string& str) {

     ^

/Users/travis/build/dremio/arrow-build/arrow/cpp/src/arrow/csv/parser_test.cc:156:6: note: candidate function

void AssertParseFinal(BlockParser& parser, const std::vector<util::string_view>& data) {

     ^
`

Comment (pitrou) [reply cpp/src/arrow/csv/parser_test.cc]: Can you try the following patch?
```diff
diff --git a/cpp/src/arrow/csv/parser_test.cc b/cpp/src/arrow/csv/parser_test.cc
index f988f3ce2..418340f54 100644
--- a/cpp/src/arrow/csv/parser_test.cc
+++ b/cpp/src/arrow/csv/parser_test.cc
@@ -135,16 +135,25 @@ Status ParseFinal(BlockParser& parser, const std::string& str, uint32_t* out_siz
   return parser.ParseFinal(util::string_view(str), out_size);
 }
 
+std::vector<util::string_view> ViewsFromStrings(const std::vector<std::string>& data) {
+  std::vector<util::string_view> views(data.size());
+  for (size_t i = 0; i < data.size(); ++i) {
+    views[i] = data[i];
+  }
+  return views;
+}
+
 void AssertParseOk(BlockParser& parser, const std::string& str) {
   uint32_t parsed_size = static_cast<uint32_t>(-1);
   ASSERT_OK(Parse(parser, str, &parsed_size));
   ASSERT_EQ(parsed_size, str.size());
 }
 
-void AssertParseOk(BlockParser& parser, const std::vector<util::string_view>& data) {
+void AssertParseOk(BlockParser& parser, const std::vector<std::string>& data) {
   uint32_t parsed_size = static_cast<uint32_t>(-1);
-  ASSERT_OK(parser.Parse(data, &parsed_size));
-  ASSERT_EQ(parsed_size, TotalViewLength(data));
+  auto views = ViewsFromStrings(data);
+  ASSERT_OK(parser.Parse(views, &parsed_size));
+  ASSERT_EQ(parsed_size, TotalViewLength(views));
 }
 
 void AssertParseFinal(BlockParser& parser, const std::string& str) {
@@ -153,10 +162,11 @@ void AssertParseFinal(BlockParser& parser, const std::string& str) {
   ASSERT_EQ(parsed_size, str.size());
 }
 
-void AssertParseFinal(BlockParser& parser, const std::vector<util::string_view>& data) {
+void AssertParseFinal(BlockParser& parser, const std::vector<std::string>& data) {
   uint32_t parsed_size = static_cast<uint32_t>(-1);
-  ASSERT_OK(parser.ParseFinal(data, &parsed_size));
-  ASSERT_EQ(parsed_size, TotalViewLength(data));
+  auto views = ViewsFromStrings(data);
+  ASSERT_OK(parser.ParseFinal(views, &parsed_size));
+  ASSERT_EQ(parsed_size, TotalViewLength(views));
 }
 
 void AssertParsePartial(BlockParser& parser, const std::string& str,
```

Comment (pprudhvi) [reply cpp/src/arrow/csv/parser_test.cc]: same error with the patch

Comment (pprudhvi) [reply cpp/src/arrow/csv/parser_test.cc]: enclosing elements of initializer list in braces seems to work. I pushed a patch - https://github.com/apache/arrow/pull/5791/files . Please merge if it looks alright. Thanks.

Comment (pitrou) [reply cpp/src/arrow/csv/parser_test.cc]: Can you open a JIRA and a separate PR? This is needed in order to track the changes to Arrow.

**Pull request closed without merging**